### PR TITLE
JSONFormatBear: Parse Exception message

### PR DESCRIPTION
**Pull request by Sanny26**
This change parses the default exception message to print
only 'Expecting property name enclosed in double quotes.'
instead of 'Expecting property name enclosed in double
quotes: line 2 column 7 (char 8)'.It also shows line and
column where the error is raised.

Closes https://github.com/coala/coala-bears/issues/35

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [ ] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [ ] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: bears/js/JSONFormatBear.py [hunk 1]
```diff
@@ -37,10 +38,16 @@ def run(self, filename, file,
             json_content = json.loads(''.join(file),
                                       object_pairs_hook=OrderedDict)
         except JSONDecodeError as err:
+            err_content = str(err).split(':')
+            err_msg = err_content[0]+'.'
+            position = findall(r'\d+', err_content[1])
```
Comment (Makman2) [inline bears/js/JSONFormatBear.py]: I think you should make a single regex and use capture groups accordingly, so your parsing just requires a single match:
```
m = match(r'(.*): line (\d+) column (\d+)'
err_msg = m.group(1) + '.'
...
    line=int(m.group(1))
    column=int(m.group(2))
```

File: bears/js/JSONFormatBear.py [hunk 2]
```diff
@@ -37,10 +38,16 @@ def run(self, filename, file,
...
             yield Result.from_values(
                 self,
-                'This file does not contain parsable JSON. ' + repr(str(err)),
-                file=filename)
+                'This file does not contain parsable JSON. ' +
+                repr(str(err_msg)),
```
Comment (Makman2) [inline bears/js/JSONFormatBear.py]: though you haven't introduced it, if you want you can push another commit which corrects this to just `err_msg` instead of `repr(str(err_msg))` (as `err_msg` is already a string, `str` has no effect + `repr` also renders the string like it has to look in a python file when you write this string, just the message itself is okay).

File: bears/js/JSONFormatBear.py [hunk 3]
```diff
@@ -37,10 +38,16 @@ def run(self, filename, file,
             json_content = json.loads(''.join(file),
                                       object_pairs_hook=OrderedDict)
         except JSONDecodeError as err:
+            err_content = str(err).split(':')
```
Comment (aptrishu) [inline bears/js/JSONFormatBear.py]: use regex and write a test for it.. to verify the output.

Comment (aptrishu) [inline bears/js/JSONFormatBear.py]: Ok, @Makman2 has pointed it out.. :+1:

**Comment by Sanny26** [comment]
@Makman2 If a user runs it over an empty file, the output it prints is 'The line belonging to the following result cannot be printed because it refers to a line that doesn't seem to exist in the given file'. 
As a user, I think just the error message 'Expecting value' is sufficient. What do you suggest? Should that message be displayed? 

And, yeah I'l change it to single expression. I was just tempted to use split because I remember reading that split is fast as compared to re.split and regex is generally slow.

File: bears/js/JSONFormatBear.py (hunk 3 above)
Comment (Sanny26) [inline bears/js/JSONFormatBear.py]: @aptrishu coverage is 100%, so is a test needed?

**Comment by Makman2** [comment]
> @Makman2 If a user runs it over an empty file, the output it prints is 'The line belonging to the following result cannot be printed because it refers to a line that doesn't seem to exist in the given file'.
As a user, I think just the error message 'Expecting value' is sufficient. What do you suggest? Should that message be displayed?

If you can catch this case it would be good. Though "Expecting value" is bad imo, I mean if I would read the message my brain would be totally confused. Rather say "file is empty", because that's the case ;)

File: bears/js/JSONFormatBear.py (hunk 3 above)
Comment (Makman2) [inline bears/js/JSONFormatBear.py]: Problem is that those tests don't check for the actual result message. If possible write some tests that check whether your regex is working correctly^^

**Comment by Sanny26** [comment]
@Makman2 @aptrishu I added the tests. Could you review them?

File: bears/js/JSONFormatBear.py [hunk 4]
```diff
@@ -33,14 +39,24 @@ def run(self, filename, file,
         :param escape_unicode: Whether or not to escape unicode values using
                                ASCII.
         """
+        if stat(filename).st_size == 0:
```
Comment (Makman2) [inline bears/js/JSONFormatBear.py]: naaah please use `file`:
`if len(file) == 0:`
\+ I would also add a little comment explaining why you explicitly catch this case
\+ I would move this into a separate commit

File: bears/js/JSONFormatBear.py [hunk 5]
```diff
@@ -33,14 +39,24 @@ def run(self, filename, file,
...
+            yield Result.from_values(self,
+                                     'This file is empty',
```
Comment (Makman2) [inline bears/js/JSONFormatBear.py]: Missing period at end

File: tests/js/JSONFormatBearTest.py [hunk 6]
```diff
@@ -25,6 +26,20 @@
...
+    def test_parse_err_msg(self):
+        valid_msg = ['Expecting value: line 20 column 30 (char 45)',
+                     'Expecting value: line 1 column 1 ']
+        invalid_msg = ['Expecting property change',
+                       'Expecting line 20 column 30',
+                       'Expecting value: line column']
+        for msg in valid_msg:
+            self.assertTrue(JSONFormatBear.parse_err_msg(msg))
+        for msg in invalid_msg:
+            self.assertFalse(JSONFormatBear.parse_err_msg(msg))
```
Comment (Makman2) [inline tests/js/JSONFormatBearTest.py]: Could you rather try to assert for the values? See https://github.com/coala/coala-bears/blob/master/tests/vcs/git/GitCommitBearTest.py for how this works^^

Comment (Makman2) [inline tests/js/JSONFormatBearTest.py]: Then you can also inline `parse_err_msg` again^^

Comment (Sanny26) [inline tests/js/JSONFormatBearTest.py]: For tests, I can assert group(1), group(2), group(3) for the expected output. Would that be sufficient?
And by inline, did you mean something like this?
```
def my_function():
    def other_function():
        pass
```

Comment (Makman2) [inline tests/js/JSONFormatBearTest.py]: inlining means in this case to inline the content, not just the function ;)

> For tests, I can assert group(1), group(2), group(3) for the expected output. Would that be sufficient?

I think you should still assert for the bear results, as they get delivered to the user, and that's where we want to test ideally.

**Comment by Asnelchristian** [comment]
@Sanny26 would you address @Makman2 issues ? :sweat_smile:

**Comment by Asnelchristian** [comment]
needs a rebase

**Comment by Sanny26** [comment]
@Asnelchristian I was busy with few evaluations....will complete by sunday.

File: bears/js/JSONFormatBear.py [hunk 7]
```diff
@@ -1,5 +1,7 @@
 import json
+from os import stat
```
Comment (sahilarora535) [inline bears/js/JSONFormatBear.py]: As @Makman2 suggested, use `len`, `stat` no required. Also, if the file has say a newline, or any other whitespace character, and is not empty, then your code will give the same error since the file would not be empty in size but actually is a `null` JSON.

**Comment by Sanny26** [comment]
@Makman2  I made the suggested changes. For tests, I used check_results function from LocalBearTestHelper. Or am I supposed to write it in a similar fashion as GitCommitBearTest only?
And, if these tests look fine, then I'll move the empty file test case part to a new commit.

File: tests/js/JSONFormatBearTest.py [hunk 8]
```diff
@@ -1,6 +1,9 @@
-from bears.js import JSONFormatBear
-from coalib.testing.LocalBearTestHelper import verify_local_bear
-
+from queue import Queue
+from bears.js.JSONFormatBear import JSONFormatBear
+from coalib.testing.LocalBearTestHelper import (verify_local_bear,
+                                                LocalBearTestHelper)
+from coalib.results.Result import Result
+from coalib.settings.Section import Section
```
Comment (Makman2) [inline tests/js/JSONFormatBearTest.py]: add a newline :)

File: tests/js/JSONFormatBearTest.py [hunk 9]
```diff
@@ -1,6 +1,9 @@
-from bears.js import JSONFormatBear
-from coalib.testing.LocalBearTestHelper import verify_local_bear
-
+from queue import Queue
```
Comment (Makman2) [inline tests/js/JSONFormatBearTest.py]: could you add a newline here to separate system imports from package imports? :)

File: tests/js/JSONFormatBearTest.py [hunk 10]
```diff
@@ -25,7 +28,40 @@
...
+        self.section = Section('')
+        self.uut = JSONFormatBear(self.section, Queue())
+
+    def test_exception_result(self):
+        self.check_results(
+            self.uut,
+            test_file4.split('\n'),
+            [Result.from_values('JSONFormatBear',
+                                '''This file does not contain parsable JSON. \
+Expecting property name enclosed in double quotes.''',
```
Comment (Makman2) [inline tests/js/JSONFormatBearTest.py]: that looks bad :)
rather use single quotes and do this:
```
Result.from_values('JSONFormatBear',
                   'This file does not contain parsable JSON. '
                       'Expecting property name enclosed in '
                       'double quotes.'
```
or
```
Result.from_values(
    'JSONFormatBear',
    'This file does not contain parsable JSON. Expecting property '
        'name enclosed in double quotes.',
    file='default',
    ...
```

**Comment by Makman2** [comment]
Exactly what I wanted to see in the tests :)

Looks good, next iteration should be mergable :+1:

**Comment by Sanny26** [comment]
@Makman2  Changes made as suggested! :)

File: bears/js/JSONFormatBear.py [hunk 11]
```diff
@@ -34,6 +34,13 @@ def run(self, filename, file,
         :param escape_unicode: Whether or not to escape unicode values using
```
Comment (Makman2) [inline bears/js/JSONFormatBear.py]: commit message: Add `JSONFormatBear` as tag :)

**Comment by Makman2** [comment]
ack 44dd843

**Comment by Makman2** [comment]
unack ce733cb

**Comment by Sanny26** [comment]
Added JSONFormatBear tag and issue reference as well.

File: bears/js/JSONFormatBear.py (hunk 11 above)
Comment (Makman2) [inline bears/js/JSONFormatBear.py]: Ah one last: Please remove the "Closes XXXX" reference in the commit message, it does not really apply :)

**Comment by Makman2** [comment]
unack a971c85

**Comment by Sanny26** [comment]
Removed issue reference as suggested.

**Comment by Makman2** [comment]
ack c4ceb3f

**Comment by Makman2** [comment]
rebase and lgtm :)

**Comment by Sanny26** [comment]
Rebase done!

**Comment by Makman2** [comment]
reack 5e7b526 462b953

**Comment by Makman2** [comment]
@rultor merge
